tfhe/src/integer/server_key/radix_parallel/sub.rs

```diff
@@ -486,4 +784,108 @@ impl ServerKey {
         self.key.create_trivial_assign(&mut borrows_out[0], 0);
         (borrows_out, last_block_out_borrow)
     }
+
+    /// Computes the subtraction of two signed numbers and returns an indicator of overflow
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use tfhe::integer::gen_keys_radix;
+    /// use tfhe::shortint::parameters::PARAM_MESSAGE_2_CARRY_2_KS_PBS;
+    ///
+    /// // We have 4 * 2 = 8 bits of message
+    /// let num_blocks = 4;
+    /// let (cks, sks) = gen_keys_radix(PARAM_MESSAGE_2_CARRY_2_KS_PBS, num_blocks);
+    ///
+    /// let msg_1 = i8::MIN;
+    /// let msg_2 = 1;
+    ///
+    /// // Encrypt two messages:
+    /// let ctxt_1 = cks.encrypt_signed(msg_1);
+    /// let ctxt_2 = cks.encrypt_signed(msg_2);
+    ///
+    /// // Compute homomorphically a subtraction
+    /// let (result, overflowed) = sks.signed_overflowing_sub_parallelized(&ctxt_1, &ctxt_2);
+    ///
+    /// // Decrypt:
+    /// let decrypted_result: i8 = cks.decrypt_signed(&result);
+    /// let decrypted_overflow = cks.decrypt_one_block(&overflowed) == 1;
+    ///
+    /// let (expected_result, expected_overflow) = msg_1.overflowing_sub(msg_2);
+    /// assert_eq!(expected_result, decrypted_result);
+    /// assert_eq!(expected_overflow, decrypted_overflow);
+    /// ```
+    pub fn signed_overflowing_sub_parallelized(
+        &self,
+        ctxt_left: &SignedRadixCiphertext,
+        ctxt_right: &SignedRadixCiphertext,
+    ) -> (SignedRadixCiphertext, Ciphertext) {
+        let mut tmp_lhs;
+        let mut tmp_rhs;
+
+        let (lhs, rhs) = match (
+            ctxt_left.block_carries_are_empty(),
+            ctxt_right.block_carries_are_empty(),
+        ) {
+            (true, true) => (ctxt_left, ctxt_right),
+            (true, false) => {
+                tmp_rhs = ctxt_right.clone();
+                self.full_propagate_parallelized(&mut tmp_rhs);
+                (ctxt_left, &tmp_rhs)
+            }
+            (false, true) => {
+                tmp_lhs = ctxt_left.clone();
+                self.full_propagate_parallelized(&mut tmp_lhs);
+                (&tmp_lhs, ctxt_right)
+            }
+            (false, false) => {
+                tmp_lhs = ctxt_left.clone();
+                tmp_rhs = ctxt_right.clone();
+                rayon::join(
+                    || self.full_propagate_parallelized(&mut tmp_lhs),
+                    || self.full_propagate_parallelized(&mut tmp_rhs),
+                );
+                (&tmp_lhs, &tmp_rhs)
+            }
+        };
+
+        self.unchecked_signed_overflowing_sub_parallelized(lhs, rhs)
+    }
+
+    pub fn signed_overflowing_sub(
```

Comment:
should this be in the parallel module ?

Reply:
I put it there because it uses `full_propagate_parallelized` and calls `unchecked_signed_overflowing_sub` which also uses rayon.

`unchecked_signed_overflowing_sub` implements the sequential propagation algorithm
that is used when the parallel version would be slower or when the parameters do not support it.

There's a bit of a naming problem because we have two 'parallel' propagation:
- One that is truly parallelized using hillis-steele prefix sum
- One that is not truly parallelized, its the sequential algorithm, but uses rayon::join
  to parallelized some message and carry extract.

And we need to have both of those.

For `signed_overflowing_sub` I can simply remove it and only keep `signed_overflowing_sub_parallelized`

Reply:
in any case I guess rayon join is never guaranteed to be parallelized, but here this signed overflowing sub does not really have the parallelized algorithm right ?

Reply:
If you do not count rayon::join as parallelism then appart from the calls to `full_propagate_parallelized` then yes `signed_overflowing_sub` does not have parallelism.

It also call `unchecked_signed_overflowing_sub` which uses `rayon::join`

Reply:
What should we do, keep `signed_overflowing_sub`, move it to non-parallel folder, and still allow use of rayon::join.

or remove `signed_overflowing_sub` (only keep `signed_overflowing_sub_parallelized`), and rename `unchecked_signed_overflowing_sub` to `unchecked_sequential_signed_overflowing_sub_parallelized` as a private fn called when parameters are not compatible with fully parallelized variant ?

Reply:
Could be moved to the sequential module without removing the rayon stuff, I think some functions already do a similar thing anyways

Reply:
I checked, and indeed there is the unsigned non parallelized variant of overflowing_sub which uses rayon, I did that for the same reasons